DOC: Update release instructions (#9278)
astropy/astropy: docs/development/releasing.rst

      $ git checkout v1.2.x

> #. Make sure that the continuous integration services (e.g., Travis) are passing

Reviewer: Perhaps
```suggestion
#. Make sure that the continuous integration services (e.g., CircleCI) are passing
```
just because of all the uncertainty in Travis' future?

Reviewer 2: What about listing multiple? I regularly point to this docs for folks looking for instructions, and in the template we provide more for travis than for circleCI, thus many projects only use that one.

Author: OK, I added both.

   for the `astropy core repository`_ branch you are going to release. You may
   also want to locally run the tests (with remote data on to ensure all of the
   tests actually run), and make sure the description in ``setup.cfg`` is reStructuredText-

>   compliant::

Author: Ops... I think it was underindented. Will fix in the commit.
```suggestion
   compliant::
```

Reviewer 3: oops I meant to make reStructureText-compliant one hyphenated word, but wasn't sure how to do that here.

Author: That is a limitation of this "suggestion" feature on GitHub. It does not do well with multi-line editing. But no matter, I'll sort this out in my next batch of edits. Thank you!

      $ conda create -n astropy_release_test_v<version> numpy
      $ source activate astropy_release_test_v<version>

>       $ pip install .[all]

Reviewer 2: this would also install astropy, from the local directory, doesn't it?

Author: No idea. I haven't use `pip` like that much yet. Should I add a `pip uninstall astropy` after? Or should I find another way to "install all the dependencies" like @astrofrog wanted in #5120 ?

Reviewer 4: A cleaner approach is to combine this with the line below:

```
pip install dist/astropy-<version>.tar.gz[all]
```

      $ git reset --hard v<version>
      $ git push upstream stable --force

> #. Build and test the Astropy wheels.  See the `wheel builder README

Reviewer 2: nope, wheel building is before pushing to pypi. there has been sometimes issues uncovered by it, so I'm uncomfortable doing anything like this after pypi release.

Author: I moved it here because of the comment of "moving the wheel stuff after the tagging". Did I move it too far down? The numbering doesn't match anymore between the original issue and the doc, so I am just guessing here.

Reviewer 2: yes, tagging, wheels and the very last step is pypi upload (and then conda-forge).

Author: OK moving it back

      Also, it might be tempting to use the ``--tags`` argument to ``git push``,
      but this should *not* be done, as it might push up some unintended tags.

.. _post-release-procedure:

> Post-Release procedures

Author: Is this where I should draw the line between "release" and "post-release"?

Reviewer 2: pypi upload? after that the release cannot be modified? But honestly I don't really know.

Author: Okay. I moved post-release section to right after `twine upload`. The significance of this section separation is that in the RC instructions, it says "do not do the stuff in post-release".

> Post-Release procedures

Reviewer 3: ```suggestion
Post-Release Procedures
```

Author: Not sure about this one, as the other existing headings only capitalized the first letter of the first word.

Reviewer 3: No worries, will edit all of the headings when I get to this sub-package!

      $ python setup.py build sdist
      $ gpg --detach-sign -a dist/astropy-<version>.tar.gz
      $ twine check dist/*
      $ twine upload dist/astropy-<version>*

> .. _post-release-procedure:

Reviewer: ```suggestion

Congratulations!  You have completed the release! Now there are just a few
clean-up tasks to finalize the process.

.. _post-release-procedure:
```

Reviewer 2: Is this sarcasm? 😏

Author: Wait, am I supposed to commit that in or not, because I did. :scream:

Reviewer 2: It wasn't intended to be sarcasm, I think, but usually the most tedious, less fun part is the cleanups, updating the contributors page and mailmaps, etc. 🙈

   bugfix or a new major version.  You may also need to update the contributor
   list on the web site if you updated the ``docs/credits.rst`` at the outset.

#. Open a PR to the astropy *master* branch to
   update the ``CHANGES.rst`` to reflect the date of the release you just
   performed and to include the new section of the changelog.  Often the easiest
   way to do this is to use ``git cherry-pick`` the changelog commit just before
   the release commit from above. If you are not sure how to do this, you might
   be better off copying-and-pasting the relevant parts of the maintenance
   branch's ``CHANGES.rst`` into master. In the same PR, you also have to
   update ``docs/whatsnew/index.rst`` and ``docs/whatsnew/X.Y.rst`` to link to
   "what's new" documentation in the released RTD branch, using the existing
   text as example.

> #. Notify the Conda Distribution and Continuous Integration maintainers

Reviewer 2: Oh, I haven't noticed this before. Remove the first sentence, and add a new point after this one: 

* "When the conda-forge packaging is ready, email the Anaconda maintainers about the release(s) so they can update the versions in the default channels."

* "Once the release(s) are available on the default conda channels, prepare the public announcement. For new major release, coordinate with the astropy coordinators, for bugfix release(s) an e-mail to the mailing lists `astropy-dev`` and ``Astropy`` is sufficient."

Author: Isn't the second point you mentioned already the last point in this section, with "Send out release announcement emails to email lists..."?

Reviewer 2: yes, that's right. it's just adding a tiny nuance that the announcement is expected to be coordinated, and not sent out by the release manager but by the coordinators.

Author: Anyway, I attempted to apply your comments to where I think is appropriate. :woman_shrugging:

   version number (with no ``v`` prefix). Once you are happy with the changes,
   click **Save**, then **Publish**.

#. Send out release annoucement emails to email lists in
   `Getting Help with Astropy <https://www.astropy.org/help.html>`_. Use the

>    previous annoucement as a template.

Author: Ops, typo! Will fold this into the next edit cycle.
```suggestion
   previous announcement as a template.
```

.. _release-procedure-beta-rc:

Modifications for a beta/release candidate release
--------------------------------------------------

> For major releases we do beta and/or release candidates to have a chance to

Author: Not sure why it was indented, so I simply un-indented the existing text. Then, added some.

Reviewer 4: Seems reasonable

catch significant bugs before the true release. If the release you are
performing is this kind of pre-release, some of the above steps need to be
modified.

The primary modifications to the release procedure are:

* When entering the new version number, instead of just removing the
  ``.dev``, enter "1.2b1" or "1.2rc1".  It is critical that you follow this
  numbering scheme (``x.yb#`` or ``x.y.zrc#``), as it will ensure the release
  is ordered "before" the main release by various automated tools, and also
  tells PyPI that this is a "pre-release".
* Do *not* do the step of adding ``.dev`` in the "back to development" stage.
  If an RC goes well, there's no need for a "dev" stage, as the same version
  will be released with only minor doc updates, and strings like "x.yrcz.dev"
  confuse some version number parsing tools.

> * Do not do step #27 (the one *after* "Update Readthedocs...") or later, as those are tasks for an actual release.

Author: I interpret the old comments as best I can... but the number might be off. :woman_shrugging:

Reviewer 4: Maybe those tasks should be in their own section so that you can refer to the section label instead?

Author: Done...

   version number (with no ``v`` prefix). Once you are happy with the changes,
   click **Save**, then **Publish**.

#. Send out release announcement emails to email lists in
   `Getting Help with Astropy <https://www.astropy.org/help.html>`_. Use the
   previous announcement as a template.

.. _release-procedure-beta-rc:

Modifications for a beta/release candidate release
--------------------------------------------------

> For major releases we do beta and/or release candidates to have a chance to

Reviewer 3: ```suggestion
For major releases, we do beta and/or release candidates to have a chance to
```

catch significant bugs before the true release. If the release you are
performing is this kind of pre-release, some of the above steps need to be
modified.

The primary modifications to the release procedure are:

* When entering the new version number, instead of just removing the
  ``.dev``, enter "1.2b1" or "1.2rc1".  It is critical that you follow this
  numbering scheme (``x.yb#`` or ``x.y.zrc#``), as it will ensure the release
  is ordered "before" the main release by various automated tools, and also

>   tells PyPI that this is a "pre-release".

Reviewer 3: ```suggestion
  tells PyPI that this is a "pre-release."
```

* Do *not* do the step of adding ``.dev`` in the "back to development" stage.

>   If an RC goes well, there's no need for a "dev" stage, as the same version

Reviewer 3: ```suggestion
  If an RC goes well, there is no need for a "dev" stage, as the same version
```

  will be released with only minor doc updates, and strings like "x.yrcz.dev"
  confuse some version number parsing tools.

> * Do not do steps in :ref:`post-release-procedure`.

Author: Please check that the modified modifications still make sense.

> Once a release candidate is available, create a new wiki page under

Reviewer 3: ```suggestion
Once a release candidate is available, create a new Wiki page under
```

`Astropy Project Wiki <https://github.com/astropy/astropy/wiki>`_ with the
title "vX.Y RC testing" (replace "X.Y" with the release number) using the
`wiki of a previous RC <https://github.com/astropy/astropy/wiki/v3.2-RC-testing>`_
as a template.

> For releases that come after release candidates, the title of the changelog

Reviewer: It's a bit awkward, but I think this goes in the "main" release instructions, right?  Because this section is for "beta/RC" but this note is for *non*-beta/RC releases

Author: I moved it up but you can tell me whether it is where you thought it is supposed to be. 🤞

#. Repeat the above steps for the astropy-helpers, using the same version series.

> #. Create an issue titled "Release final vX.Y".

Reviewer 2: does this issue needed?

Author: @astrofrog wanted it. 🤷‍♀

Reviewer 2: oh, sure. but that was 3 years ago and he didn't complained since of not doing it. I'm OK either way, but feel it's superfluous, we never forgot to release even without having an issue.

Author: Tom R, what say you?

Reviewer 4: Remove it! 🔥

> #. Ping people to test it under some unusual cases like 32-bit Linux systems.

Reviewer 2: does this needed? the release candidate announcement was usually picked up by the right people (debian and st packaging)

Author: Just following check box written by @astrofrog .

Reviewer 4: Yeah I don't think this is needed anymore as the Debian folk follow our list.

#. Repeat the above steps for the astropy-helpers, using the same version series.

> #. Run the tests twice in a row without cleaning anything up in between.

Author: I removed 32-bit testing from here, but you still want to keep this, right?

Reviewer 2: sure 🤷‍♀ 

(confession: most of the time I don't run them at this point any more as they take forever and in "CI we trust" 🙈 )

Author: @astrofrog , you requested this. What say you?

Author: @astrofrog and I agreed that this can be removed from release instruction and added to CI directly at a later time.

Backporting fixes from master
-----------------------------

.. note::

    The changelog script in ``astropy-procedures`` does not know about

>     minor releases, thus please becareful. For example, if PRs milestoned

Reviewer 4: typo: becareful

Backporting fixes from master
-----------------------------

.. note::

>     The changelog script in ``astropy-procedures`` does not know about

Author: I removed all mentions of using `add_to_changelog.py` since @bsipocz that the script does not work anyway. Does that mean I should also remove this note, or is this referring to a different "changelog script"?

Reviewer 2: no, this comment is about the scripts in `pr_consistency`, they only check for a PR being backported to a branch, rather than checking for the milestones, too. As the comment says we have two branches: 1) master 2) 1.2.x. The release 1.2.0 will come out of 1.2.x, just as well as 1.2.1. If a PR is merged into master with 1.2.1 it shouldn't be backported until 1.2.0 is out, but the script will complain that it needs to be backported to 1.2.x nevertheless.
